### PR TITLE
enable authentication by using api key in place of password

### PR DESCRIPTION
**Pull request by nickpell**
Product Note: Allows users with two factor authentication to use Tableau/PowerBI Odata feeds by authenticating using their API key instead of their password.

Flag: https://www.commcarehq.org/hq/flags/edit/odata/

### CODE REVIEW

File: corehq/apps/api/resources/auth.py [hunk 1]
```diff
@@ -105,6 +106,19 @@ def is_authenticated(self, request, **kwargs):
         return self._auth_test(request, wrappers=wrappers, **kwargs)
 
 
+class RequirePermissionAuthenticationForOdata(RequirePermissionAuthentication):
```
Comment (esoergel) [inline corehq/apps/api/resources/auth.py]: nit: I'd probably just go with `ODataAuthentication`

**Review by esoergel — APPROVED** [review]
Shouldn't this just use `api_auth`?  Or is it important to exclude digest and session auth for some reason?

File: corehq/apps/api/resources/auth.py (hunk 1 above)
Comment (nickpell) [reply corehq/apps/api/resources/auth.py]: https://github.com/dimagi/commcare-hq/pull/24237

**Comment by nickpell** [comment]
> Shouldn't this just use api_auth? Or is it important to exclude digest and session auth for some reason?

Excluding for now because YAGNI.  I expect this feed to be accessed almost exclusively via the BI tools, which use Basic auth.  The important difference here is ```basic_or_api_key```, which extends the existing Basic auth functionality and allows for skipping the two factor check if an api key is used instead of a password.

**Comment by czue** [comment]
didn't review code but only thought was that it might be good to document how to use this somewhere (possibly in the feature flag docs?)

**Pull request merged**